### PR TITLE
PortForwardExample Added

### LINKED ISSUE #1557: How to use port-forward feature?
Hi, I'm trying to figure out how to use port-forwarding. Are there some samples of how to use it? Also - how do I manage connections? Is this done for me, or I need to call something (and when)? 

Many thanks for the great tool!

### PR DESCRIPTION
**Pull request by kunal-kushwaha**
Fixes #1557
Example added for PortForward.

### CODE REVIEW

**Comment by oscerd** [comment]
Ok to test

File: kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+  public static void main(String args[]) {
+    String master = "https://localhost:8443";
+
+    Config config = new ConfigBuilder().withMasterUrl(master).build();
+    try (final KubernetesClient client = new DefaultKubernetesClient(config)) {
+      String namespace = "default";
+      log("namespace", namespace);
+      Pod pod = client.pods().inNamespace(namespace).load(PortForwardExample.class.getResourceAsStream("/portforward-example-pod.yml")).get();
+      log("Pod created");
```
Comment (dev-gaur) [inline kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: shouldn't this be logged after the next line ? after the client has created the resource in the desired namespace ?

@rohanKanojia @kunal-kushwaha

Comment (rohanKanojia) [reply kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: yeah, makes sense

File: kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+      int containerPort =  pod.getSpec().getContainers().get(0).getPorts().get(0).getContainerPort();
+
+      LocalPortForward portForward = client.pods().withName("testpod").portForward(containerPort, 8080);
+      log("Port forwarded");
+
+      int localPort = portForward.getLocalPort();
+      SocketChannel channel = SocketChannel.open();
+      channel.connect(new InetSocketAddress("localhost", localPort));
```
Comment (rohanKanojia) [inline kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: Were you able to run this? I tried but facing this error:
```
13:23:05.999 [io.fabric8.kubernetes.examples.PortForwardExample.main()] INFO  io.fabric8.kubernetes.examples.PortForwardExample - Port forwarded
13:23:06.001 [io.fabric8.kubernetes.examples.PortForwardExample.main()] INFO  io.fabric8.kubernetes.examples.PortForwardExample - localhost/127.0.0.1:8080
13:23:06.008 [pool-2-thread-1] ERROR io.fabric8.kubernetes.client.dsl.internal.PortForwarderWebsocket - FWD: Throwable received from websocket
java.io.InterruptedIOException: executor rejected
	at okhttp3.RealCall$AsyncCall.executeOn(RealCall.java:185) ~[okhttp-3.12.0.jar:?]
	at okhttp3.Dispatcher.promoteAndExecute(Dispatcher.java:186) ~[okhttp-3.12.0.jar:?]
	at okhttp3.Dispatcher.enqueue(Dispatcher.java:137) ~[okhttp-3.12.0.jar:?]
	at okhttp3.RealCall.enqueue(RealCall.java:126) ~[okhttp-3.12.0.jar:?]
	at okhttp3.internal.ws.RealWebSocket.connect(RealWebSocket.java:193) ~[okhttp-3.12.0.jar:?]
	at okhttp3.OkHttpClient.newWebSocket(OkHttpClient.java:435) ~[okhttp-3.12.0.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.PortForwarderWebsocket.forward(PortForwarderWebsocket.java:191) ~[kubernetes-client-4.3-SNAPSHOT.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.PortForwarderWebsocket.lambda$forward$0(PortForwarderWebsocket.java:162) ~[kubernetes-client-4.3-SNAPSHOT.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.util.concurrent.RejectedExecutionException: Task okhttp3.RealCall$AsyncCall@b285a1d rejected from java.util.concurrent.ThreadPoolExecutor@7d4395fe[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2055) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:825) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1355) ~[?:?]
	at okhttp3.Rea
```

Comment (kunal-kushwaha) [reply kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: Yes, it is working for me

Comment (rohanKanojia) [reply kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: Could you please tell me steps which you followed?

Comment (kunal-kushwaha) [reply kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: I ran it on this project: 
https://github.com/kunal-kushwaha/client-demo/blob/master/src/main/java/PortForwardExample.java

Comment (rohanKanojia) [reply kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: Umm, I don't see any significant difference between these two files. What's different here?

Comment (kunal-kushwaha) [reply kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: There's no difference, I ran the same file.
I just included fabric dependencies in this demo project that I used to test examples.
And ran the same file I pushed on the repo

Comment (kunal-kushwaha) [reply kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: @rohanKanojia 
Is the issue resolved? How do I reproduce this error?

Comment (rohanKanojia) [reply kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: ah, I didn't get a chance to jump on this again. Will reproduce it soon and report back

Comment (rohanKanojia) [reply kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: Try running file using exec-maven-plugin

Comment (kunal-kushwaha) [reply kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: @rohanKanojia: Are you still facing this error? I was unable to reproduce it.

Comment (rohanKanojia) [reply kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/PortForwardExample.java]: I will take a look again here.

**Comment by oscerd** [comment]
What's the status of this?

**Comment by kunal-kushwaha** [comment]
@oscerd: It's under review. 
@rohanKanojia: Are we good to merge this? If not, do let me know if any changes are required.

**Comment by rohanKanojia** [comment]
retest this please

**Comment by oscerd** [comment]
LGTM

**Comment by rohanKanojia** [comment]
@kunal-kushwaha : Your example seems to be correct but unfortunately I have not been able to see desired behavior. I think maybe this port-forwarding feature is broken itself. 

This is port forwarding done via `kubectl`.
```
~/go/src/github.com/fabric8io/kubernetes-client : $ kubectl port-forward testpod 8081:80
Forwarding from 127.0.0.1:8081 -> 80
Forwarding from [::1]:8081 -> 80
Handling connection for 8081
Handling connection for 8081
```

After port forwarding, you should be able to access your pod from a separate terminal.
```
~/go/src/github.com/fabric8io/kubernetes-client : $ curl localhost:8081
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>

```

**Comment by rohanKanojia** [comment]
I think I need to check what's causing this error repeatedly:
```
	... 10 more
17:35:26.221 [pool-2-thread-1] ERROR io.fabric8.kubernetes.client.dsl.internal.PortForwarderWebsocket - FWD: Throwable received from websocket
java.io.InterruptedIOException: executor rejected
	at okhttp3.RealCall$AsyncCall.executeOn(RealCall.java:185) ~[okhttp-3.12.0.jar:?]
	at okhttp3.Dispatcher.promoteAndExecute(Dispatcher.java:186) ~[okhttp-3.12.0.jar:?]
	at okhttp3.Dispatcher.enqueue(Dispatcher.java:137) ~[okhttp-3.12.0.jar:?]
	at okhttp3.RealCall.enqueue(RealCall.java:126) ~[okhttp-3.12.0.jar:?]
	at okhttp3.internal.ws.RealWebSocket.connect(RealWebSocket.java:193) ~[okhttp-3.12.0.jar:?]
	at okhttp3.OkHttpClient.newWebSocket(OkHttpClient.java:435) ~[okhttp-3.12.0.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.PortForwarderWebsocket.forward(PortForwarderWebsocket.java:191) ~[classes/:?]
	at io.fabric8.kubernetes.client.dsl.internal.PortForwarderWebsocket.lambda$forward$0(PortForwarderWebsocket.java:162) ~[classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.util.concurrent.RejectedExecutionException: Task okhttp3.RealCall$AsyncCall@73f8aa58 rejected from java.util.concurrent.ThreadPoolExecutor@54058d31[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2055) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:825) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1355) ~[?:?]
	at okhttp3.RealCall$AsyncCall.executeOn(RealCall.java:182) ~[okhttp-3.12.0.jar:?]
	... 10 more
```

No worries, I would take care of this :-)

**Comment by kunal-kushwaha** [comment]
@rohanKanojia: Thank you for your help. I really appreciate it.

**Comment by rohanKanojia** [comment]
[merge]

**Pull request merged**